src/library/scala/collection/immutable/Vector.scala

```diff
@@ -219,6 +220,7 @@ final class Vector[+A] private[immutable] (private[collection] val startIndex: I
     s.dirty = dirty
     s.gotoPosWritable(focus, idx, focus ^ idx)  // if dirty commit changes; go to new pos and prepare for writing
     s.display0(idx & 31) = elem.asInstanceOf[AnyRef]
+    ScalaRunTime.releaseFence()
```

Comment:
Echoing @adriaanm's comments from #7028 that belong over in this PR.

> is this to publish the mutation of dirty?

Yes, that's the intent. 

>  Should we also do this in the constructor?

Good catch, I hadn't considered that. I think we get away without one because it starts with the default value. I'll add a comment to that effect. Also happy to add the fence if that seems to subtle.

Reply:
On second thoughts, I'll add the fence. We need the memory barrier to safely publish the assignments to the _other_ `val` fields in Vector, and the presence of a single non-final field means the implicit fence at the end of the constructor is not guaranteed (although it is added in practice for classes with >= 1 final fields)